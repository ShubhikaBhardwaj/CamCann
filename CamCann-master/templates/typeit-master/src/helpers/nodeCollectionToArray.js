/**
 * Convert a NodeList to an array of nodes. I only need this because
 * some browsers do not yet support Array.from().
 *
 * @param {NodeList} nodeList
 * @return {array}
 */
export default nodeList => {
  return [].slice.call(nodeList);
};
