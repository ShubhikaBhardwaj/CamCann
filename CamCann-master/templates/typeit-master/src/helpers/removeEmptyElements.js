import removeNode from "./removeNode";
import nodeCollectionToArray from "./nodeCollectionToArray";

/**
 * Check if a given character node is empty, either containing nothing or an empty HTML element.
 *
 * @param {object} node
 * @return {boolean}
 */
export const characterIsEmpty = node => {
  // It's a text node. Leave it be.
  if (node.nodeType === 3) {
    return false;
  }

  // Break tags are an exception.
  if (node.tagName === "BR") {
    return false;
  }

  // If there's no first child, this is empty!
  return !node.firstChild;
};

/**
 * Check if array of nodes contains any empty characters.
 *
 * @param {array} nodes
 * @return {boolean}
 */
export const containsEmptyCharacter = nodes => {
  return nodes.some(node => {
    return characterIsEmpty(node);
  });
};

/**
 * Given a DOM scope and selector, remove any HTML element remnants.
 *
 * @param {object} scope
 * @param {string} selector
 * @return {void}
 */
export default node => {
  let allHTMLNodes = nodeCollectionToArray(node.querySelectorAll("*"));
  let hasEmptyNodes = containsEmptyCharacter(allHTMLNodes);

  while (allHTMLNodes.length && hasEmptyNodes) {
    let shouldRequery = false;

    allHTMLNodes.forEach(char => {
      if (characterIsEmpty(char)) {
        removeNode(char);
        shouldRequery = true;
      }
    });

    // Re-query, since we just removed nodes.
    // Conditionally do this, to avoid unnecessary queries.
    if (shouldRequery) {
      allHTMLNodes = nodeCollectionToArray(node.querySelectorAll("*"));
    }

    // Removing nodes might have created new empty nodes,
    // so we need to re-query and check again.
    hasEmptyNodes = containsEmptyCharacter(allHTMLNodes);
  }

  return allHTMLNodes;
};
