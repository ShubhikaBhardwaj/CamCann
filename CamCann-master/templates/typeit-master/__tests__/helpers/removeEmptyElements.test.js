import nodeCollectionToArray from "../../src/helpers/nodeCollectionToArray";
import removeEmptyElements, {
  characterIsEmpty,
  containsEmptyCharacter
} from "../../src/helpers/removeEmptyElements";

describe("characterIsEmpty()", () => {
  test("Returns false when a break tag is encountered.", () => {
    let result = characterIsEmpty(document.createElement("BR"));
    expect(result).toBe(false);
  });

  test("Returns false when character has text.", () => {
    let node = document.createTextNode("hi.");
    let result = characterIsEmpty(node);
    expect(result).toBe(false);
  });

  test("Returns true when character has empty HTML", () => {
    setHTML`<span></span>`;
    let node = document.querySelector("span");
    let result = characterIsEmpty(node);
    expect(result).toBe(true);
  });

  test("Returns false when character contains HTML containing other character (empty or not).", () => {
    setHTML`<span><a></a></span>`;
    let node = document.querySelector("span");
    let result = characterIsEmpty(node);
    expect(result).toBe(false);
  });
});

describe("containsEmptyCharacter()", () => {
  test("Returns true when at least one character is empty.", () => {
    setHTML`<span>hi.<span></span>again.</span>`;
    let nodes = nodeCollectionToArray(
      document.querySelector("span").childNodes
    );
    let result = containsEmptyCharacter(nodes);
    expect(result).toBe(true);
  });

  test("Returns false when none are empty.", () => {
    setHTML`sup.hi.`;
    let nodes = nodeCollectionToArray(document.body.childNodes);
    let result = containsEmptyCharacter(nodes);
    expect(result).toBe(false);
  });
});

describe("removeEmptyElements()", () => {
  test("Empty HTML character should be removed from DOM.", () => {
    setHTML`<span id="scope">a<span></span>bc</span>`;
    removeEmptyElements(document.getElementById("scope"));
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  test("Nested empty HTML character should be removed from DOM.", () => {
    setHTML`<span id="scope">ab<span><a></a></span>c</span>`;

    removeEmptyElements(document.getElementById("scope"));
    expect(document.body.innerHTML).toMatchSnapshot();
  });
});
