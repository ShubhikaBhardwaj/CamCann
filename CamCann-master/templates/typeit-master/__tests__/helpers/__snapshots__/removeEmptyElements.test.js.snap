// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`removeEmptyElements() Empty HTML character should be removed from DOM. 1`] = `"<span id=\\"scope\\">abc</span>"`;

exports[`removeEmptyElements() Nested empty HTML character should be removed from DOM. 1`] = `"<span id=\\"scope\\">abc</span>"`;
